sklearn/tree/_tree.pxd

```diff
@@ -31,6 +31,15 @@ cdef struct Node:
     unsigned char missing_go_to_left     # Whether features have missing values
 
 
+cdef struct ParentInfo:
```

Comment:
We would have only `n_constant_features` and `impurity`, I think I would not do the change. But the fact that to build a tree we need more info from the parent, I kind of like this structure. It makes the code more readable and we can document in the structure why do we need these statistics for.

The semantic makes much more sense to me than the alternative PR and the size of the tree structure remain the same.

One thing that we should pay attention however is if we potentially have a performance regression because of a potential indirection when accessing the variable in the structure. I don't recall as well if we should pay attention to struct padding or paccking when adding those (or the compiler will take care for us).

Reply:
> We would have only n_constant_features and impurity, I think I would not do the change. 

There is also the lower/upper bound and possibly other "stuff" in the future. 

> But the fact that to build a tree we need more info from the parent, I kind of like this structure. It makes the code more readable and we can document in the structure why do we need these statistics for. The semantic makes much more sense to me than the alternative PR and the size of the tree structure remain the same.

Agreed! 

> One thing that we should pay attention however is if we potentially have a performance regression because of a potential indirection when accessing the variable in the structure. 

Agreed as well! I can look into this. My initial glance of the code would hypothesize this is not immediately a huge issue because of where we dereference the `ParentInfo` struct. It is not occurring within a loop mostly. 

I can run some exps comparing n_samples vs n_dims for 100, and 500 trees.

Reply:
> There is also the lower/upper bound and possibly other "stuff" in the future.

Whoops, I forgot to start my sentence with "if" :). Basically, I'm happy with the change.

Reply:
I've done a bit of experimentation and reading up on struct padding just to be sure. There is really no observable difference because:

i) our structs are not that complex,
ii) a struct of float64_t and intptr_t have at most padding done on intptr_t I think, but they are the same size except on 32-bit machines, where there will be padding of 32-bits added to the intptr_t.
iii) we have 1 ParentInfo passed around per tree, so unless we're training thousands/millions of trees in parallel, we will not see any observable memory diff.

See some of the basic exps ran in the PR description. Overall I also looped in the `lower_bound` and `upper_bound` into the `ParentRecord` struct. Lmk wdyt @glemaitre ?